lib/recomb_map.c

```diff
@@ -196,6 +208,50 @@ recomb_map_phys_to_discrete_genetic(recomb_map_t *self, double x, uint32_t *locu
     return ret;
 }
 
+/* Find the `index` of the interval within `values` the `query` fits, such that
+ * values[index-1] < query <= values[index]
+ * If necessary, will find the leftmost such index
+ */
+size_t
+recomb_map_binary_interval_search(double query, double *values, size_t n_values)
+{
+    if (n_values == 0) {
+        return 0;
+    }
+
+    size_t l = 0;
+    size_t r = n_values;
+    size_t m, index;
+
+
+    while (l < r) {
+        m = (size_t)(floor((double)(l + r) / 2L));
+
+        if (values[m] == query) {
+            index = m;
+            goto fin;
+        } else if ((values[m] < query) && (m != l)) {
+            l = m;
+        } else if ((values[m] > query) && (m != r)) {
+            r = m;
+        } else {
+            l ++;
+            r = l;
+        }
```

Comment:
Could you avoid the linear search by adjusting this logic to something like `values[m] < query => l = m + 1` and otherwise `r = m`? You also wouldn't need the boundary checks since the `while` loop should catch that.

I'm new to the group so am poking around some issues/PRs (so please correct me if I'm missing anything)!

Reply:
I am also new here!

The suggestion seems to work as far as not needing to take the branch above (237-240). However, I still need the ugly iteration [248-251](https://github.com/ivan-krukov/msprime/blob/recomb_map/lib/recomb_map.c#L248), since I want the leftmost value on the recombination map (I think, I might be missing the point). The tests affected are `test_zero_rate_two_intervals (tests.test_recombination_map.TestCoordinateConversion)` [here](https://github.com/ivan-krukov/msprime/blob/recomb_map/tests/test_recombination_map.py#L180).

Reply:
Ah, part of my suggestion included removing the if statement that checks equality between the value at `m` and `query`, leaving just the less than and else cases. This allows `l` and `r` to converge onto the leftmost equivalent element as opposed to stopping when any one is found which would require further iteration (this is the same as finding the "insertion point" of `query`).

A small additional note: you can remove the need for the if statements before and after the loop by initializing `r = n_values - 1`.

Reply:
I see! This is great, it really simplifies things. Not sure why I settled on the verbose version in retrospect. 

Thank you.